homeassistant/components/syncthru/config_flow.py

```diff
@@ -0,0 +1,128 @@
+        )
+        if any(x.data[CONF_URL] == url for x in self._async_current_entries()):
+            return self.async_abort(reason="already_configured")
+
+        name = discovery_info.get(ssdp.ATTR_UPNP_FRIENDLY_NAME)
+        if name:
+            # Remove trailing " (ip)" if present for consistency with user driven config
+            name = re.sub(r"\s+\([\d.]+\)\s*$", "", name)
+
+        # https://github.com/PyCQA/pylint/issues/3167
+        self.context.update(  # pylint: disable=no-member
+            {CONF_URL: url, CONF_NAME: name, "title_placeholders": {CONF_NAME: name}}
+        )
+        return await self.async_step_confirm()
+
+    async def async_step_confirm(self, user_input=None):
+        """Handle discovery confirmation by user."""
+        if user_input is not None:
+            return await self._async_check_and_create("confirm", user_input)
+
+        return await self._async_show_form(
+            step_id="confirm",
+            user_input={
+                # https://github.com/PyCQA/pylint/issues/3167
+                CONF_URL: self.context[CONF_URL],  # pylint: disable=no-member
+                CONF_NAME: self.context[CONF_NAME],  # pylint: disable=no-member
+            },
+        )
+
+    async def _async_show_form(self, step_id, user_input=None, errors=None):
+        """Show our form."""
+        if user_input is None:
+            user_input = {}
+        return self.async_show_form(
+            step_id=step_id,
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_URL, default=user_input.get(CONF_URL, "")): str,
+                    vol.Optional(CONF_NAME, default=user_input.get(CONF_NAME, "")): str,
+                }
+            ),
+            errors=errors or {},
+        )
+
+    async def _async_check_and_create(self, step_id, user_input):
+        """Validate input, proceed to create."""
+        user_input[CONF_URL] = url_normalize(
+            user_input[CONF_URL], default_scheme="http"
+        )
+        if "://" not in user_input[CONF_URL]:
+            return await self._async_show_form(
+                step_id=step_id, user_input=user_input, errors={CONF_URL: "invalid_url"}
+            )
+
+        await self.async_set_unique_id(user_input[CONF_URL])
+        self._abort_if_unique_id_configured()
+
+        if any(
+            x.data[CONF_URL] == user_input[CONF_URL]
+            for x in self._async_current_entries()
+        ):
+            return self.async_abort(reason="already_configured")
+
+        session = aiohttp_client.async_get_clientsession(self.hass)
+        printer = SyncThru(user_input[CONF_URL], session)
+        errors = {}
+        try:
+            await printer.update()
+            if not user_input.get(CONF_NAME):
+                user_input[CONF_NAME] = DEFAULT_NAME_TEMPLATE.format(
+                    printer.model() or DEFAULT_MODEL
+                )
+        except ValueError:
```

Comment:
Can you catch a more specific error ? 

Can you also distinguish when you get a wrong url and are unable to connect?

Reply:
Re more specific error: unfortunately not, `ValueError` is the thing documented to be raised on unsupported models:
https://github.com/nielstron/pysyncthru/blob/394fdac15e2534c3439bee1bf218784426006dce/pysyncthru/__init__.py#L37

Reply:
Re wrong URL: we could detect that by examining the artificial status generated by pysyncthru in this case:
https://github.com/nielstron/pysyncthru/blob/394fdac15e2534c3439bee1bf218784426006dce/pysyncthru/__init__.py#L47

However I'm not entirely sure if we should, pysyncthru seems to want to treat these as "offline" cases and not as errors. Also I'm not sure what would be the desired way to handle a case where something that was configured using YAML before is offline at time the config entry import runs. For other cases it would be fine to report an error, but for the imports I suppose we should just let it pass. Thoughts?

Reply:
> Re more specific error: unfortunately not, ValueError is the thing documented to be raised on unsupported models, and also implemented that way:

Good point, it might be a good idea to introduce a more specific error than "ValueError" (i.e. "NoSyncthruSupport (ValueError)" and raise it.

> However I'm not entirely sure if we should, pysyncthru seems to want to treat these as "offline" cases and not as errors. 
The idea is the following: It might be that a printer is not turned on all the time (i.e. for saving energy) and still I would not want an error to be thrown in home assistant. If this use case is too specific, it might be removed within the library but anyways you can also just check whether the state of the printer is "offline" from within the ha script.

Reply:
Actually the code already handles the unable to connect case, it does it through checking `printer.is_unknown_state()` which I gather is a good approach here.